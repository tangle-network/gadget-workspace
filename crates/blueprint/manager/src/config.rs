use clap::Parser;
use std::path::PathBuf;

#[derive(Debug, Parser)]
#[command(
    name = "Blueprint Manager",
    about = "An program executor that connects to the Tangle network and runs protocols dynamically on the fly"
)]
pub struct BlueprintManagerConfig {
    /// The path to the gadget configuration file
    #[arg(short = 's', long)]
    pub gadget_config: Option<PathBuf>,
    /// The path to the keystore
    #[arg(short = 'k', long)]
    pub keystore_uri: String,
    /// The directory in which all gadgets will store their data
    #[arg(long, short = 'd', default_value = "./data")]
    pub data_dir: PathBuf,
    /// The verbosity level, can be used multiple times to increase verbosity
    #[arg(long, short = 'v', action = clap::ArgAction::Count)]
    pub verbose: u8,
    /// Whether to use pretty logging
    #[arg(long)]
    pub pretty: bool,
    /// An optional unique string identifier for the blueprint manager to differentiate between multiple
    /// running instances of a BlueprintManager (mostly for debugging purposes)
    #[arg(long, alias = "id")]
    pub instance_id: Option<String>,
    #[arg(long, short = 't')]
    pub test_mode: bool,
}
