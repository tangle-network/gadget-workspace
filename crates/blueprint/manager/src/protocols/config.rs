use gadget_std::collections::HashMap;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct ProtocolConfig {
    pub(crate) protocols: Vec<ProtocolToml>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProtocolToml {
    pub package: String,
    pub repository: Option<HashMap<String, String>>,
    pub bin_hashes: Option<HashMap<String, String>>,
}
