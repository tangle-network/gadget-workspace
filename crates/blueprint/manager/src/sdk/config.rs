use gadget_keystore::KeystoreConfig;
use libp2p::Multiaddr;
use std::net::IpAddr;
use std::path::PathBuf;
use tangle_subxt::tangle_testnet_runtime::api::runtime_types::tangle_primitives::services::ServiceBlueprint;

#[derive(Debug)]
pub struct SingleGadgetConfig {
    pub base_path: PathBuf,
    pub keystore: KeystoreConfig,
    pub bind_ip: IpAddr,
    pub bind_port: u16,
    pub bootnodes: Vec<Multiaddr>,
    pub services: Vec<ServiceBlueprint>,
    pub n_protocols: usize,
}
