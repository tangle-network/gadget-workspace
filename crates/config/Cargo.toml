[package]
name = "gadget-config"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true

[dependencies]
# Core dependencies (always included)
clap = { workspace = true, features = ["derive", "env"] }
gadget-std = { workspace = true }
serde = { workspace = true, features = ["derive"] }
serde_json = { workspace = true, features = ["alloc"], optional = true }
thiserror = { workspace = true }
tracing = { workspace = true }
url = { workspace = true, features = ["serde"] }

# Optional dependencies (feature-gated)
libp2p = { workspace = true, optional = true }
gadget-networking = { workspace = true, optional = true }
gadget-keystore = { workspace = true, optional = true, features = ["sp-core", "std"] }
alloy-primitives = { workspace = true, features = ["serde"], optional = true }

[dev-dependencies]
alloy-primitives = { workspace = true, features = ["serde"] }
gadget-config = { path = ".", features = ["std", "eigenlayer", "symbiotic", "tangle", "networking"] }
gadget-std = { workspace = true, features = ["std"] }
libp2p = { workspace = true }


[features]
default = ["std"]
std = [
	"alloy-primitives/std",
	"clap/std",
	"gadget-std/std",
	"serde/std",
	"serde_json/std",
]
eigenlayer = ["dep:alloy-primitives"]
symbiotic = ["dep:alloy-primitives"]
tangle = []
networking = ["libp2p", "gadget-networking", "keystore", "gadget-keystore"]

# Core feature groups
keystore = []

# Testing features
test-utils = ["std"]
