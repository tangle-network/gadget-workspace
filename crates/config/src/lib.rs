#![allow(unused_variables, unreachable_code)]
use gadget_std::fmt::Debug;
use gadget_std::string::{String, ToString};

#[cfg(feature = "std")]
use gadget_std::path::PathBuf;
#[cfg(not(feature = "std"))]
pub type PathBuf = String;

pub mod context_config;
pub mod protocol;
pub mod supported_chains;

pub use context_config::{ContextConfig, GadgetCLICoreSettings};
pub use protocol::{Protocol, ProtocolSettings};

/// Errors that can occur while loading and using the gadget configuration.
#[derive(Debug, thiserror::Error)]
#[non_exhaustive]
pub enum Error {
    /// Missing `RPC_URL` environment variable.
    #[error("Missing Tangle RPC endpoint")]
    MissingTangleRpcEndpoint,
    /// Missing `KEYSTORE_URI` environment
    #[error("Missing keystore URI")]
    MissingKeystoreUri,
    /// Missing `BLUEPRINT_ID` environment variable
    #[error("Missing blueprint ID")]
    MissingBlueprintId,
    /// Missing `SERVICE_ID` environment variable
    #[error("Missing service ID")]
    MissingServiceId,
    /// Error parsing the blueprint ID.
    #[error(transparent)]
    MalformedBlueprintId(core::num::ParseIntError),
    /// Error parsing the service ID.
    #[error(transparent)]
    MalformedServiceId(core::num::ParseIntError),
    /// Unsupported keystore URI.
    #[error("Unsupported keystore URI: {0}")]
    UnsupportedKeystoreUri(String),
    /// Error parsing the protocol, from the `PROTOCOL` environment variable.
    #[error("Unsupported protocol: {0}")]
    UnsupportedProtocol(String),
    /// No Sr25519 keypair found in the keystore.
    #[error("No Sr25519 keypair found in the keystore")]
    NoSr25519Keypair,
    /// Invalid Sr25519 keypair found in the keystore.
    #[error("Invalid Sr25519 keypair found in the keystore")]
    InvalidSr25519Keypair,
    /// No ECDSA keypair found in the keystore.
    #[error("No ECDSA keypair found in the keystore")]
    NoEcdsaKeypair,
    /// Invalid ECDSA keypair found in the keystore.
    #[error("Invalid ECDSA keypair found in the keystore")]
    InvalidEcdsaKeypair,
    /// Test setup error
    #[error("Test setup error: {0}")]
    TestSetup(String),
    /// Missing `EigenlayerContractAddresses`
    #[error("Missing EigenlayerContractAddresses")]
    MissingEigenlayerContractAddresses,
    /// Missing `SymbioticContractAddresses`
    #[error("Missing SymbioticContractAddresses")]
    MissingSymbioticContractAddresses,
    #[error("Bad RPC Connection: {0}")]
    BadRpcConnection(String),
    #[error("Configuration error: {0}")]
    ConfigurationError(String),
}

#[cfg(feature = "networking")]
use libp2p::Multiaddr;

pub type StdGadgetConfiguration = GadgetConfiguration;

/// Gadget environment.
#[non_exhaustive]
#[derive(Debug, Clone, Default)]
pub struct GadgetConfiguration {
    /// HTTP RPC endpoint for host restaking network (Tangle / Ethereum (Eigenlayer or Symbiotic)).
    pub http_rpc_endpoint: String,
    /// WS RPC endpoint for host restaking network (Tangle / Ethereum (Eigenlayer or Symbiotic)).
    pub ws_rpc_endpoint: String,

    /// The keystore URI for the gadget
    pub keystore_uri: String,
    /// Data directory exclusively for this gadget
    ///
    /// This will be `None` if the blueprint manager was not provided a base directory.
    pub data_dir: Option<PathBuf>,
    /// The list of bootnodes to connect to
    #[cfg(feature = "networking")]
    pub bootnodes: Vec<Multiaddr>,
    /// The type of protocol the gadget is executing on.
    pub protocol: Protocol,
    /// Protocol-specific settings
    pub protocol_settings: ProtocolSettings,
    /// Whether the gadget is in test mode
    pub test_mode: bool,
}

/// Loads the [`GadgetConfiguration`] from the current environment.
/// # Errors
///
/// This function will return an error if any of the required environment variables are missing.
pub fn load(config: ContextConfig) -> Result<GadgetConfiguration, Error> {
    load_inner(config)
}

fn load_inner(config: ContextConfig) -> Result<GadgetConfiguration, Error> {
    tracing::info_span!("gadget");
    let ContextConfig {
        gadget_core_settings:
            GadgetCLICoreSettings::Run {
                test_mode,
                http_rpc_url,
                ws_rpc_url,
                #[cfg(feature = "networking")]
                bootnodes,
                keystore_uri,
                protocol,
                #[cfg(feature = "tangle")]
                blueprint_id,
                #[cfg(feature = "tangle")]
                service_id,
                #[cfg(feature = "eigenlayer")]
                registry_coordinator,
                #[cfg(feature = "eigenlayer")]
                operator_state_retriever,
                #[cfg(feature = "eigenlayer")]
                delegation_manager,
                #[cfg(feature = "eigenlayer")]
                service_manager,
                #[cfg(feature = "eigenlayer")]
                stake_registry,
                #[cfg(feature = "eigenlayer")]
                strategy_manager,
                #[cfg(feature = "eigenlayer")]
                avs_directory,
                #[cfg(feature = "eigenlayer")]
                rewards_coordinator,
                #[cfg(feature = "symbiotic")]
                operator_registry,
                #[cfg(feature = "symbiotic")]
                network_registry,
                #[cfg(feature = "symbiotic")]
                base_delegator,
                #[cfg(feature = "symbiotic")]
                network_opt_in_service,
                #[cfg(feature = "symbiotic")]
                vault_opt_in_service,
                #[cfg(feature = "symbiotic")]
                slasher,
                #[cfg(feature = "symbiotic")]
                veto_slasher,
                ..
            },
        ..
    } = config;

    let protocol_settings = if cfg!(feature = "tangle") && matches!(protocol, Protocol::Tangle) {
        #[cfg(feature = "tangle")]
        {
            ProtocolSettings::from_tangle(crate::protocol::TangleInstanceSettings {
                blueprint_id: blueprint_id.ok_or(Error::MissingBlueprintId)?,
                service_id: Some(service_id.ok_or(Error::MissingServiceId)?),
            })
        }
        #[cfg(not(feature = "tangle"))]
        {
            return Err(Error::UnsupportedProtocol("tangle".to_string()));
        }
    } else if cfg!(feature = "eigenlayer") && matches!(protocol, Protocol::Eigenlayer) {
        #[cfg(feature = "eigenlayer")]
        {
            ProtocolSettings::from_eigenlayer(crate::protocol::EigenlayerContractAddresses {
                registry_coordinator_address: registry_coordinator
                    .ok_or(Error::MissingEigenlayerContractAddresses)?,
                operator_state_retriever_address: operator_state_retriever
                    .ok_or(Error::MissingEigenlayerContractAddresses)?,
                delegation_manager_address: delegation_manager
                    .ok_or(Error::MissingEigenlayerContractAddresses)?,
                service_manager_address: service_manager
                    .ok_or(Error::MissingEigenlayerContractAddresses)?,
                stake_registry_address: stake_registry
                    .ok_or(Error::MissingEigenlayerContractAddresses)?,
                strategy_manager_address: strategy_manager
                    .ok_or(Error::MissingEigenlayerContractAddresses)?,
                avs_directory_address: avs_directory
                    .ok_or(Error::MissingEigenlayerContractAddresses)?,
                rewards_coordinator_address: rewards_coordinator
                    .ok_or(Error::MissingEigenlayerContractAddresses)?,
            })
        }
        #[cfg(not(feature = "eigenlayer"))]
        {
            return Err(Error::UnsupportedProtocol("eigenlayer".to_string()));
        }
    } else if cfg!(feature = "symbiotic") && matches!(protocol, Protocol::Symbiotic) {
        #[cfg(feature = "symbiotic")]
        {
            ProtocolSettings::from_symbiotic(crate::protocol::SymbioticContractAddresses {
                operator_registry_address: operator_registry
                    .ok_or(Error::MissingSymbioticContractAddresses)?,
                network_registry_address: network_registry
                    .ok_or(Error::MissingSymbioticContractAddresses)?,
                base_delegator_address: base_delegator
                    .ok_or(Error::MissingSymbioticContractAddresses)?,
                network_opt_in_service_address: network_opt_in_service
                    .ok_or(Error::MissingSymbioticContractAddresses)?,
                vault_opt_in_service_address: vault_opt_in_service
                    .ok_or(Error::MissingSymbioticContractAddresses)?,
                slasher_address: slasher.ok_or(Error::MissingSymbioticContractAddresses)?,
                veto_slasher_address: veto_slasher
                    .ok_or(Error::MissingSymbioticContractAddresses)?,
            })
        }
        #[cfg(not(feature = "symbiotic"))]
        {
            return Err(Error::UnsupportedProtocol("symbiotic".to_string()));
        }
    } else {
        return Err(Error::UnsupportedProtocol(protocol.to_string()));
    };

    Ok(GadgetConfiguration {
        test_mode,
        http_rpc_endpoint: http_rpc_url.to_string(),
        ws_rpc_endpoint: ws_rpc_url.to_string(),
        keystore_uri,
        #[cfg(feature = "std")]
        data_dir: gadget_std::env::var("DATA_DIR").ok().map(PathBuf::from),
        #[cfg(not(feature = "std"))]
        data_dir: None,
        #[cfg(feature = "networking")]
        bootnodes: bootnodes.unwrap_or_default(),
        protocol,
        protocol_settings,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::protocol::{Protocol, ProtocolSettings};

    #[test]
    fn verify_cli() {
        use clap::CommandFactory;
        ContextConfig::command().debug_assert();
    }

    #[test]
    fn test_default_configuration() {
        let config = GadgetConfiguration::default();
        assert!(config.http_rpc_endpoint.is_empty());
        assert!(config.ws_rpc_endpoint.is_empty());
        assert!(config.keystore_uri.is_empty());
        assert!(config.data_dir.is_none());
        assert!(!config.test_mode);
        assert_eq!(config.protocol, Protocol::default());
    }

    // Test Eigenlayer configuration when feature is enabled
    #[cfg(feature = "eigenlayer")]
    mod eigenlayer_tests {
        use super::*;
        use crate::protocol::EigenlayerContractAddresses;

        #[test]
        fn test_eigenlayer_configuration() {
            let addresses = EigenlayerContractAddresses::default();

            let settings = ProtocolSettings::from_eigenlayer(addresses);
            assert!(matches!(settings, ProtocolSettings::Eigenlayer(_)));
            if let ProtocolSettings::Eigenlayer(addrs) = settings {
                assert_eq!(
                    addrs.registry_coordinator_address,
                    addresses.registry_coordinator_address
                );
                assert_eq!(
                    addrs.operator_state_retriever_address,
                    addresses.operator_state_retriever_address
                );
                assert_eq!(
                    addrs.delegation_manager_address,
                    addresses.delegation_manager_address
                );
                assert_eq!(
                    addrs.service_manager_address,
                    addresses.service_manager_address
                );
                assert_eq!(
                    addrs.stake_registry_address,
                    addresses.stake_registry_address
                );
                assert_eq!(
                    addrs.strategy_manager_address,
                    addresses.strategy_manager_address
                );
                assert_eq!(addrs.avs_directory_address, addresses.avs_directory_address);
            }
        }
    }

    // Test Symbiotic configuration when feature is enabled
    #[cfg(feature = "symbiotic")]
    mod symbiotic_tests {
        use alloy_primitives::address;

        use super::*;
        use crate::protocol::SymbioticContractAddresses;

        #[test]
        fn test_symbiotic_configuration() {
            let addresses = SymbioticContractAddresses {
                operator_registry_address: address!("8431e038b88ba5945ce8bf41e4af1374f3fb6e4c"),
                network_registry_address: address!("4f4495243837681061c4743b74b3eedf548d56a5"),
                base_delegator_address: address!("2279b7a0a67db372996a5fab50d91eaa73d2ebe6"),
                network_opt_in_service_address: address!(
                    "610178da211fef7d417bc0e6fed39f05609ad788"
                ),
                vault_opt_in_service_address: address!("8a791620dd6260079bf849dc5567adc3f2fdc318"),
                slasher_address: address!("f9e5b5c6a4b5b9c1d4e8f7a3b2e1d0c9a8b7f6e5"),
                veto_slasher_address: address!("a1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8a9b0"),
            };

            let settings = ProtocolSettings::from_symbiotic(addresses);
            assert!(matches!(settings, ProtocolSettings::Symbiotic(_)));
            if let ProtocolSettings::Symbiotic(addrs) = settings {
                assert_eq!(
                    addrs.operator_registry_address,
                    addresses.operator_registry_address
                );
                assert_eq!(
                    addrs.network_registry_address,
                    addresses.network_registry_address
                );
                assert_eq!(
                    addrs.base_delegator_address,
                    addresses.base_delegator_address
                );
                assert_eq!(
                    addrs.network_opt_in_service_address,
                    addresses.network_opt_in_service_address
                );
                assert_eq!(
                    addrs.vault_opt_in_service_address,
                    addresses.vault_opt_in_service_address
                );
                assert_eq!(addrs.slasher_address, addresses.slasher_address);
                assert_eq!(addrs.veto_slasher_address, addresses.veto_slasher_address);
            }
        }
    }

    #[test]
    fn test_configuration_validation() {
        // Test RPC endpoint validation
        let config = GadgetConfiguration {
            http_rpc_endpoint: "invalid-url".to_string(),
            ..Default::default()
        };
        assert!(validate_rpc_endpoint(&config.http_rpc_endpoint).is_err());

        // Test keystore URI validation
        let config = GadgetConfiguration {
            keystore_uri: "invalid-uri".to_string(),
            ..Default::default()
        };
        assert!(validate_keystore_uri(&config.keystore_uri).is_err());
    }

    // Helper functions for tests
    fn validate_rpc_endpoint(endpoint: &str) -> Result<(), Error> {
        if !endpoint.starts_with("http://") && !endpoint.starts_with("https://") {
            return Err(Error::BadRpcConnection(
                "Invalid RPC URL format".to_string(),
            ));
        }
        Ok(())
    }

    fn validate_keystore_uri(uri: &str) -> Result<(), Error> {
        if !uri.starts_with("file://") {
            return Err(Error::UnsupportedKeystoreUri(uri.to_string()));
        }
        Ok(())
    }
}
