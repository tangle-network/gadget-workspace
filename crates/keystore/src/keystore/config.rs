/// The config for a [`Keystore`]
///
/// Depending on the features enabled, this provides methods to enable different storage backends.
///
/// ### Implicit registration
///
/// When no other backends are enabled, an [`InMemoryStorage`] will implicitly be registered by [`Keystore::new()`].
///
/// This means that:
///
/// ```rust
/// use gadget_keystore::{Keystore, KeystoreConfig};
///
/// # fn main() -> gadget_keystore::Result<()> {
/// let config = KeystoreConfig::new().in_memory(true);
/// let keystore = Keystore::new(config)?;
/// # Ok(()) }
/// ```
///
/// is equivalent to:
///
/// ```rust
/// use gadget_keystore::{Keystore, KeystoreConfig};
///
/// # fn main() -> gadget_keystore::Result<()> {
/// let keystore = Keystore::new(KeystoreConfig::new())?;
/// # Ok(()) }
/// ```
///
/// [`InMemoryStorage`]: crate::storage::InMemoryStorage
/// [`Keystore`]: crate::Keystore
/// [`Keystore::new()`]: crate::Keystore::new
#[derive(Default, Debug)]
pub struct KeystoreConfig {
    pub(crate) in_memory: bool,
    #[cfg(feature = "std")]
    pub(crate) fs_root: Option<std::path::PathBuf>,
    #[cfg(any(
        feature = "aws-signer",
        feature = "gcp-signer",
        feature = "ledger-browser",
        feature = "ledger-node"
    ))]
    pub(crate) remote_configs: Vec<crate::remote::RemoteConfig>,
}

impl KeystoreConfig {
    /// Create a new `KeystoreConfig`
    ///
    /// Alias for [`KeystoreConfig::default()`]
    ///
    /// # Examples
    ///
    /// ```rust
    /// use gadget_keystore::{Keystore, KeystoreConfig};
    ///
    /// # fn main() -> gadget_keystore::Result<()> {
    /// let config = KeystoreConfig::new();
    /// let keystore = Keystore::new(config)?;
    /// # Ok(()) }
    /// ```
    pub fn new() -> Self {
        Self::default()
    }

    /// Register an [`InMemoryStorage`] backend
    ///
    /// NOTE: This will be enabled by default if no other storage backends are enabled.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use gadget_keystore::{Keystore, KeystoreConfig};
    ///
    /// # fn main() -> gadget_keystore::Result<()> {
    /// let config = KeystoreConfig::new().in_memory(true);
    /// let keystore = Keystore::new(config)?;
    /// # Ok(()) }
    /// ```
    ///
    /// [`InMemoryStorage`]: crate::storage::InMemoryStorage
    pub fn in_memory(mut self, value: bool) -> Self {
        self.in_memory = value;
        self
    }

    /// Register a [`FileStorage`] backend
    ///
    /// See [`FileStorage::new()`] for notes on how `path` is used.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// use gadget_keystore::{Keystore, KeystoreConfig};
    ///
    /// # fn main() -> gadget_keystore::Result<()> {
    /// let config = KeystoreConfig::new().fs_root("path/to/keystore");
    /// let keystore = Keystore::new(config)?;
    /// # Ok(()) }
    /// ```
    ///
    /// [`FileStorage`]: crate::storage::FileStorage
    /// [`FileStorage::new()`]: crate::storage::FileStorage::new
    #[cfg(feature = "std")]
    pub fn fs_root<P: AsRef<std::path::Path>>(mut self, path: P) -> Self {
        self.fs_root = Some(path.as_ref().to_path_buf());
        self
    }

    cfg_remote! {
        /// Register a remote backend
        ///
        /// See [`RemoteConfig`] for available options.
        ///
        /// # Examples
        ///
        /// ```rust,no_run
        /// use gadget_keystore::{Keystore, KeystoreConfig};
        /// use gadget_keystore::remote::RemoteConfig;
        ///
        /// # fn main() -> gadget_keystore::Result<()> {
        /// let remote = RemoteConfig::Aws {
        ///     keys: vec![]
        /// };
        ///
        /// let config = KeystoreConfig::new().remote(remote);
        /// let keystore = Keystore::new(config)?;
        /// # Ok(()) }
        /// ```
        ///
        /// [`RemoteConfig`]: crate::remote::RemoteConfig
        pub fn remote(mut self, remote_config: crate::remote::RemoteConfig) -> Self {
            self.remote_configs.push(remote_config);
            self
        }
    }

    fn is_empty(&self) -> bool {
        let mut is_empty = self.in_memory;
        #[cfg(feature = "std")]
        {
            is_empty |= self.fs_root.is_none();
        }
        #[cfg(any(
            feature = "aws-signer",
            feature = "gcp-signer",
            feature = "ledger-browser",
            feature = "ledger-node"
        ))]
        {
            is_empty |= !self.remote_configs.is_empty();
        }

        is_empty
    }

    pub(crate) fn finalize(self) -> Self {
        if self.is_empty() {
            return self.in_memory(true);
        }

        self
    }
}
