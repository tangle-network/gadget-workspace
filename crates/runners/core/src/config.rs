use crate::error::RunnerError;
use gadget_config::GadgetConfiguration;

#[async_trait::async_trait]
pub trait BlueprintConfig: Send + Sync + 'static {
    async fn register(
        &self,
        _env: &GadgetConfiguration,
    ) -> Result<(), RunnerError> {
        Ok(())
    }
    async fn requires_registration(
        &self,
        _env: &GadgetConfiguration,
    ) -> Result<bool, RunnerError> {
        Ok(true)
    }
}

impl BlueprintConfig for () {}