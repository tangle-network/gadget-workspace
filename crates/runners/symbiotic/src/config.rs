use crate::{
    config::{GadgetConfiguration, ProtocolSpecificSettings},
    error, info,
    keystore::BackendExt,
    utils::evm::{get_provider_http, get_wallet_provider_http},
};
use alloy_network::EthereumWallet;
use symbiotic_rs::OperatorRegistry;

use super::{BlueprintConfig, RunnerError};

#[derive(Clone, Copy, Default)]
pub struct SymbioticConfig {}

#[async_trait::async_trait]
impl BlueprintConfig for SymbioticConfig {
    async fn requires_registration(
        &self,
        env: &GadgetConfiguration<parking_lot::RawRwLock>,
    ) -> Result<bool, RunnerError> {
        if env.skip_registration {
            return Ok(false);
        }

        let ProtocolSpecificSettings::Symbiotic(contract_addresses) = &env.protocol_specific else {
            return Err(RunnerError::InvalidProtocol(
                "Expected Symbiotic protocol".into(),
            ));
        };
        let operator_registry_address = contract_addresses.operator_registry_address;

        let operator_address = env.keystore()?.ecdsa_key()?.alloy_key()?.address();
        let operator_registry = OperatorRegistry::new(
            operator_registry_address,
            get_provider_http(&env.http_rpc_endpoint),
        );

        let is_registered = operator_registry
            .isEntity(operator_address)
            .call()
            .await
            .map(|r| r._0)
            .map_err(|e| RunnerError::SymbioticError(e.to_string()))?;

        Ok(!is_registered)
    }

    async fn register(
        &self,
        env: &GadgetConfiguration<parking_lot::RawRwLock>,
    ) -> Result<(), RunnerError> {
        let ProtocolSpecificSettings::Symbiotic(contract_addresses) = &env.protocol_specific else {
            return Err(RunnerError::InvalidProtocol(
                "Expected Symbiotic protocol".into(),
            ));
        };
        let operator_registry_address = contract_addresses.operator_registry_address;

        let operator_signer = env.keystore()?.ecdsa_key()?.alloy_key()?;
        let wallet = EthereumWallet::new(operator_signer);
        let provider = get_wallet_provider_http(&env.http_rpc_endpoint, wallet);
        let operator_registry = OperatorRegistry::new(operator_registry_address, provider.clone());

        let result = operator_registry
            .registerOperator()
            .send()
            .await?
            .get_receipt()
            .await?;

        if result.status() {
            info!("Operator registered successfully");
        } else {
            error!("Operator registration failed");
        }

        Ok(())
    }
}
