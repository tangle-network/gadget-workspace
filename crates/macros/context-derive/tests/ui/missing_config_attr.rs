use blueprint_sdk::config::GadgetConfiguration;
use gadget_context_derive::KeystoreContext;

#[derive(KeystoreContext)]
struct MyContext {
    foo: String,
    sdk_config: GadgetConfiguration,
}

fn main() {}
