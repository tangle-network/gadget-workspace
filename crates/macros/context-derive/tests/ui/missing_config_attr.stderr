error: No field with #[config] attribute found, please add #[config] to the field that holds the `gadget_sdk::config::GadgetConfiguration`
 --> tests/ui/missing_config_attr.rs:5:8
  |
5 | struct MyContext {
  |        ^^^^^^^^^
